pydm/main_window.py

```diff
@@ -107,166 +105,80 @@ def set_display_widget(self, new_widget):
-        try:
-            if not os.path.exists(filename):
-                new_fname = find_display_in_path(ui_file)
-                if new_fname is None or new_fname == "":
-                    raise IOError("File {} not found".format(filename))
-                filename = new_fname
-            self.new_abs_window(filename, macros, command_line_args)
-        except (IOError, OSError, ValueError, ImportError) as e:
-            error_msg = "Cannot open file: '{0}'. Reason: '{1}'.".format(filename, e)
-            logger.error(error_msg)
-            self.statusBar().showMessage(error_msg, 5000)
-
-    def new_abs_window(self, filename, macros=None, command_line_args=None):
-        merged_macros = self.merge_with_current_macros(macros)
-        self.app.new_window(filename, merged_macros, command_line_args)
-
-    def go_button_pressed(self):
-        filename = str(self.ui.panelSearchLineEdit.text())
-        if not filename:
-            return
-        try:
-            if QApplication.keyboardModifiers() == Qt.ShiftModifier:
-                self.app.new_window(filename)
-            else:
-                self.go(filename)
-        except (IOError, OSError, ValueError, ImportError) as e:
-            self.handle_open_file_error(filename, e)
-
     def handle_open_file_error(self, filename, error):
         self.statusBar().showMessage("Cannot open file: '{0}', reason: '{1}'...".format(filename, error), 5000)
 
-    # Note: in go(), back(), and forward(), always do history stack manipulation *before* opening the file.
-    # That way, the navigation button enable/disable state will work correctly.  This is stupid, and will be fixed eventually.
-    def go(self, ui_file, macros=None, command_line_args=None):
-        self.forward_stack = []
-        self.open_file(ui_file, macros, command_line_args)
+    def back(self):
+        curr_display = self.display_widget()
+        prev_display = None
+        if curr_display:
+            prev_display = curr_display.previous_display()
 
-    def go_abs(self, ui_file, macros=None, command_line_args=None):
-        self.forward_stack = []
-        self.open_abs_file(filename=ui_file, macros=macros, command_line_args=command_line_args)
+        if not prev_display:
+            logger.error('No display history to execute back navigation.')
+            return
 
-    def back(self):
-        if len(self.back_stack) > 1:
-            if QApplication.keyboardModifiers() == Qt.ShiftModifier:
-                stack_item = self.back_stack[-2]
-                self.new_abs_window(filename=stack_item[0], macros=stack_item[1], command_line_args=stack_item[2])
-            else:
-                self.forward_stack.append(self.back_stack.pop())
-                stack_item = self.back_stack[-1]
-                self.open_abs_file(filename=stack_item[0], macros=stack_item[1], command_line_args=stack_item[2])
+        if QApplication.keyboardModifiers() == Qt.ShiftModifier:
```

Comment:
From the standpoint of someone who might use these methods (back, forward, home) in code rather than just as a pyqtSignal callback, it'd be advisable to keep the functionality tied to a method argument and not whether the user is pressing shift.

That is, make it something like: `def back(self, open_in_new_process: bool = False)` and call it with `.back(open_in_new_process=(QApplication.keyboardModifiers() == Qt.ShiftModifier))`

Reply:
Addressed via 62acf85